cpp/src/arrow/flight/types.h

```diff
@@ -109,10 +115,37 @@ struct Ticket {
   std::string ticket;
 };
 
-/// \brief A host location (hostname and port)
+class FlightClient;
+
+/// \brief A host location (a URI)
 struct Location {
-  std::string host;
-  int32_t port;
+ public:
+  /// \brief Initialize a blank location.
+  Location();
+  /// \brief Initialize a location by parsing a URI string
+  static Status Parse(const std::string& uri_string, Location& location);
```

Comment:
We use pointers for out-parameters, so `Location* location`. Following the Google C++ style guide: https://google.github.io/styleguide/cppguide.html#Output_Parameters